app/controllers/availabilities_controller.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+class AvailabilitiesController < ApplicationController
+  before_action :authenticate_user!
+
+  def index
+    @time_ranges = TimeRange.order(:start_time)
+    @availabilities = load_availabilities
+    @start_time = start_time
+    @end_time = end_time
+  end
+
+  def update_availabilities
+    availability_ids = params[:availability_ids] || []
+    Availability.where(user_id: current_user.id).each do |availability|
+      availability.update(status:
+        availability_ids.include?(availability.id.to_s))
+    end
+    redirect_to availabilities_path
+  end
+
+  private
+
+  def load_availabilities
```

Comment:
Notice that this method calls `O(mn)` SQL queries where `m` is the number of time ranges and `n` is the number of days!

Ideally you only want `O(1)` queries. Try starting from `Availability.where(user: current_user).joins(:time_range).order('day', 'time_ranges.start_time')`.

Reply:
Apart from that you can also see that the `TimeRange` call gets called a lot as well. This one I leave it to you.

1. If you notice, at the top left there's a small button. You can check the performance of the page there. Find which line is causing it.
2. Generally, these queries are caused due to using a non-ActiveRecord data structure (i.e. arrays etc). The idea is, you want to call everything once in a possibly big SQL queries, then you iterate through that. That way you only use 1 database trip.
3. Have a look at ActiveRecord eager loading (`#include`).

Let me know if you're stuck. :)

Reply:
Reworked, now only query if necessary. Now queries are `O(1)`. `update_availabilities` worst case still `O(mn)`, but only update if it changes.

Reply:
Every row update is one SQL query so O(mn) is ok. Dont think you can batch update